### PR TITLE
Add contract results support to acceptance tests

### LINKED ISSUE #2981: INSUFFICIENT_GAS during acceptance test
### Description

The mirror node acceptance tests execute a contract that consistently hits an `INSUFFICIENT_GAS` error response from the main nodes. Even though we set a max gas limit to 1M it fails with gas used of 300K. Turns out main nodes currently have a max gas limit of 300K and we're hitting that limit and failing as a result. We need to simplify our contract to use less storage/gas until smart contracts 2.0 is released with increased limits. Below is an example:

https://app.dragonglass.me/hedera/transactions/009501639168131509506419

### Steps to reproduce

Execute mirror node acceptance tests that include contract tests.

### Additional context

```
com.hedera.mirror.test.e2e.acceptance.steps.ContractFeature.verifyDeployedContractMirror()
2021-12-10T14:29:02.291-0600 DEBUG main o.s.r.s.RetryTemplate Retry: count=0 
2021-12-10T14:29:02.445-0600 DEBUG main o.s.r.s.RetryTemplate Retry: count=0 
2021-12-10T14:29:02.594-0600 WARN main c.h.h.s.TransactionReceiptQuery Retrying node 0.0.3 in 250 ms after failure during attempt #1: OK 
2021-12-10T14:29:02.993-0600 WARN main c.h.h.s.TransactionReceiptQuery Retrying node 0.0.3 in 500 ms after failure during attempt #2: OK 
2021-12-10T14:29:03.643-0600 WARN main c.h.h.s.TransactionReceiptQuery Retrying node 0.0.3 in 1000 ms after failure during attempt #3: OK 
2021-12-10T14:29:04.793-0600 WARN main c.h.h.s.TransactionReceiptQuery Retrying node 0.0.3 in 2000 ms after failure during attempt #4: OK 
2021-12-10T14:29:06.934-0600 WARN main c.h.h.s.TransactionReceiptQuery Retrying node 0.0.3 in 4000 ms after failure during attempt #5: OK 
2021-12-10T14:29:16.076-0600 DEBUG main o.s.r.s.RetryTemplate Checking for rethrow: count=1 
2021-12-10T14:29:16.077-0600 DEBUG main o.s.r.s.RetryTemplate Retry: count=1 
2021-12-10T14:29:16.214-0600 DEBUG main o.s.r.s.RetryTemplate Checking for rethrow: count=2 
2021-12-10T14:29:16.215-0600 DEBUG main o.s.r.s.RetryTemplate Retry failed last attempt: count=2 
2021-12-10T14:29:16.226-0600 ERROR main c.h.m.t.e.a.c.ContractClient TransactionReceipt retrieval Exception: receipt for transaction 0.0.950@1639168131.509506419 raised status INSUFFICIENT_GAS 
2021-12-10T14:29:16.365-0600 DEBUG main o.s.r.s.RetryTemplate Retry: count=0 
2021-12-10T14:29:21.502-0600 DEBUG main o.s.r.s.RetryTemplate Checking for rethrow: count=1 
2021-12-10T14:29:21.503-0600 DEBUG main o.s.r.s.RetryTemplate Retry: count=1 
2021-12-10T14:29:21.641-0600 DEBUG main o.s.r.s.RetryTemplate Checking for rethrow: count=2 
2021-12-10T14:29:21.641-0600 DEBUG main o.s.r.s.RetryTemplate Retry failed last attempt: count=2 
  When I successfully call the contract                                                                 # com.hedera.mirror.test.e2e.acceptance.steps.ContractFeature.callContract()
      com.hedera.hashgraph.sdk.ReceiptStatusException: receipt for transaction 0.0.950@1639168131.509506419 raised status INSUFFICIENT_GAS
	at com.hedera.hashgraph.sdk.TransactionId.getReceipt(TransactionId.java:122)
	at com.hedera.hashgraph.sdk.TransactionId.getRecord(TransactionId.java:144)
	at com.hedera.mirror.test.e2e.acceptance.client.AbstractNetworkClient.lambda$getTransactionRecord$2(AbstractNetworkClient.java:132)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209)
	at com.hedera.mirror.test.e2e.acceptance.client.AbstractNetworkClient.getTransactionRecord(AbstractNetworkClient.java:132)
	at com.hedera.mirror.test.e2e.acceptance.client.ContractClient.executeContract(ContractClient.java:141)
	at com.hedera.mirror.test.e2e.acceptance.steps.ContractFeature.executeCreateChildTransaction(ContractFeature.java:195)
	at com.hedera.mirror.test.e2e.acceptance.steps.ContractFeature.callContract(ContractFeature.java:89)
	at ???.I successfully call the contract(file:///home/jenkins/agent/workspace/hcs-acceptance/hedera-mirror-test/src/test/resources/features/contract/contract.feature:12)
```

### Hedera network

mainnet

### Version

v0.45.1

### Operating system

_No response_

### PR DESCRIPTION
**Pull request by Nana-EC**
**Description**:
The current contract support in the acceptance test has the following issues
- Insufficient gas supply error since London release of Ethereum
- No contract results coverage

To update support
- Simplify Parent/Child smart contract to reduce gas used
- Update Hardhat smart contract tests
- Add contract results verification for contact results by id and transactionId 

**Related issue(s)**:

Fixes #2981 

**Notes for reviewer**:
Ran `@contractbase` and `@acceptance` against integration to confirm

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java [hunk 1]
```diff
@@ -121,6 +123,17 @@ public MirrorContractResponse getContractInfo(String contractId) {
         return callRestEndpoint("/contracts/{contractId}", MirrorContractResponse.class, contractId);
     }
 
+    public MirrorContractResultsResponse getContractResultsById(String contractId) {
+        log.debug("Verify contract results'{}' is returned by Mirror Node", contractId);
+        return callRestEndpoint("/contracts/{contractId}/results", MirrorContractResultsResponse.class, contractId);
+    }
+
+    public MirrorContractResultResponse getContractResultsByTransactionId(String transactionId) {
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java:131]:
```suggestion
    public MirrorContractResultResponse getContractResultByTransactionId(String transactionId) {
```

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java [hunk 2]
```diff
@@ -38,4 +38,11 @@ public static Instant messageQueryDateStringToInstant(String date, Instant refer
 
         return refDate;
     }
+
+    public static String solidityAddress(long shard, long realm, long num) {
+        return String.format("0x%s%s%s",
+                String.format("%08X", shard),
+                String.format("%016X", realm),
+                String.format("%016X", num)).toLowerCase();
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java:46]:
```suggestion
        return String.format("0x%08x%016x%016x", shard, realm, num)
```

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java [hunk 3]
```diff
@@ -191,6 +208,67 @@ private MirrorContractResponse verifyContractFromMirror(boolean isDeleted) throw
         return mirrorContract;
     }
 
+    private void verifyContractExecutionResultsById() {
+        List<MirrorContractResult> contractResults = mirrorClient.getContractResultsById(contractId.toString())
+                .getResults();
+
+        assertNotNull(contractResults);
+        assertThat(contractResults).isNotEmpty();
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java:216]: `isNotEmpty()` also checks for not null.

```suggestion
        assertThat(contractResults).isNotEmpty();
```

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java [hunk 4]
```diff
@@ -191,6 +208,67 @@ private MirrorContractResponse verifyContractFromMirror(boolean isDeleted) throw
...
+
+        contractResults.forEach(result -> {
+            verifyContractExecutionResults(result);
+        });
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java:220]:
```suggestion
        assertThat(contractResults).isNotEmpty().allSatisfy(this::verifyContractExecutionResult);
```

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java [hunk 5]
```diff
@@ -191,6 +208,67 @@ private MirrorContractResponse verifyContractFromMirror(boolean isDeleted) throw
...
+        ContractExecutionStage contractExecutionStage = contractResult.getFunctionParameters().equals("0x") ?
+                ContractExecutionStage.CREATION : ContractExecutionStage.CALL;
+
+        assertThat(contractResult.getCallResult()).isNotBlank();
+        assertThat(contractResult.getContractId()).isEqualTo(contractId.toString());
+        String[] createdIds = contractResult.getCreatedContractIds();
+        assertThat(createdIds).isNotEmpty();
+        assertThat(contractResult.getErrorMessage()).isBlank();
+        assertThat(contractResult.getFrom()).isEqualTo(FeatureInputHandler.solidityAddress(
+                contractClient.getSdkClient().getExpandedOperatorAccountId().getAccountId().shard,
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java:243]: Could add overloaded ContractID and AccountID methods for `solidityAddress()` to simplify this.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java:243]: Yeah. Sucks IDs don't all inherit from a base EntityID, but that's a protobuf limitation.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java [hunk 6]
```diff
@@ -191,6 +208,67 @@ private MirrorContractResponse verifyContractFromMirror(boolean isDeleted) throw
...
+                contractClient.getSdkClient().getExpandedOperatorAccountId().getAccountId().realm,
+                contractClient.getSdkClient().getExpandedOperatorAccountId().getAccountId().num));
```
Comment (ijungmann) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java:245]: Store the accountId as a local variable to not call this extended get three times.

File: hedera-mirror-test/src/test/resources/solidity/README.md [hunk 7]
```diff
@@ -1,7 +1,8 @@
 # Hedera Mirror Node Sample Smart Contract Project
 
-This project demonstrates a basic smart contract use case using the [Hardhat](https://hardhat.org/) ethereum development environment. 
-It comes with a sample contract, a test for that contract, and a a sample script that deploys that contract, and an example of a task implementation, which simply lists the available accounts.
+This project demonstrates a basic smart contract use case using the [Hardhat](https://hardhat.org/) ethereum development
+environment. It comes with a sample contract, a test for that contract, and a sample script that deploys that contract,
+and an example of a task implementation, which simply lists the available accounts.
```
Comment (ijungmann) [inline hedera-mirror-test/src/test/resources/solidity/README.md:5]:
```suggestion
environment. It comes with a sample contract, a test for that contract, a sample script that deploys that contract,
and an example of a task implementation which simply lists the available accounts.
```

**Review by ijungmann — CHANGES_REQUESTED** [review]
Couple small nits

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java:245]: Yeah, weighed the pros of getter vs the extra variable. For test code your suggestion is better.
Have refactored this though so only need to call it once.

File: hedera-mirror-test/src/test/resources/solidity/README.md (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/resources/solidity/README.md]: Updated

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java]: Updated

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java]: Updated

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java]: Updated

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java]: Baha, iterated on so many version I missed that. 
Updated

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: Updated

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java [hunk 8]
```diff
@@ -39,10 +42,15 @@ public static Instant messageQueryDateStringToInstant(String date, Instant refer
...
+    public static String solidityAddress(ContractId contractId) {
+        return FeatureInputHandler.solidityAddress(contractId.shard, contractId.realm, contractId.num);
+    }
+
     public static String solidityAddress(long shard, long realm, long num) {
-        return String.format("0x%s%s%s",
-                String.format("%08X", shard),
-                String.format("%016X", realm),
-                String.format("%016X", num)).toLowerCase();
+        return String.format("0x%08x%016x%016x", shard, realm, num).toLowerCase();
```
Comment (steven-sheehy) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java:54]: nit: The `toLowerCase()` is not needed with the use of `x` vs `X`

**Review by steven-sheehy — APPROVED** [review]

**Review by ijungmann — DISMISSED** [review]

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java [hunk 9]
```diff
@@ -121,6 +123,17 @@ public MirrorContractResponse getContractInfo(String contractId) {
         return callRestEndpoint("/contracts/{contractId}", MirrorContractResponse.class, contractId);
     }
 
+    public MirrorContractResultsResponse getContractResultsById(String contractId) {
+        log.debug("Verify contract results'{}' is returned by Mirror Node", contractId);
+        return callRestEndpoint("/contracts/{contractId}/results", MirrorContractResultsResponse.class, contractId);
+    }
+
+    public MirrorContractResultResponse getContractResultByTransactionId(String transactionId) {
+        log.debug("Verify contract results'{}' is returned by Mirror Node", transactionId);
```
Comment (ijungmann) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java:132]: nit:
```suggestion
        log.debug("Verify contract result'{}' is returned by Mirror Node", transactionId);
```

Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java]: Fixed

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java (hunk 8 above)
Comment (Nana-EC) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/util/FeatureInputHandler.java]: Removed

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**